src/OauthTokenFileStorage.php

```diff
@@ -116,85 +160,93 @@ public function markExpired(): void {
    * {@inheritdoc}
    */
   public function saveToken(array $data): void {
+    // Even if an error occurs here token data can be still served from the
+    // internal cache in this page request.
+    try {
+      $this->checkRequirements();
+    }
+    catch (OauthTokenStorageException $exception) {
+      $this->logger->critical('OAuth token file storage: %error.', ['%error' => $exception->getMessage()]);
+    }
     // Calculate the cache expiration.
     $data['expires'] = isset($data['expires_in']) ? $data['expires_in'] + time() : ($data['expires'] ?? -1);
     // Remove the expires_in data.
     unset($data['expires_in']);
 
-    // Gets the file directory so we can make sure it exists.
-    $file_path = dirname(static::oauthTokenPath());
-    file_prepare_directory($file_path, FILE_CREATE_DIRECTORY | FILE_MODIFY_PERMISSIONS);
-
     // Write the obfuscated token data to a private file.
-    file_unmanaged_save_data(base64_encode(serialize($data)), static::oauthTokenPath(), FILE_EXISTS_REPLACE);
+    file_unmanaged_save_data(base64_encode(serialize($data)), $this->tokenFilePath, FILE_EXISTS_REPLACE);
 
     // Update the cached value.
-    static::$token_data = $data;
+    $this->tokenData = $data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function checkRequirements(): void {
+    if (strpos($this->tokenFilePath, 'private://') === 0 && empty($this->settings->get('file_private_path'))) {
+      throw new OauthTokenStorageException('Unable to save token data to private filesystem because it has not been configured yet.');
+    }
+    // Gets the file directory so we can make sure it exists.
+    $token_directory = dirname($this->tokenFilePath);
+    if (!file_prepare_directory($token_directory, FILE_CREATE_DIRECTORY | FILE_MODIFY_PERMISSIONS)) {
+      throw new OauthTokenStorageException("Unable to set up {$token_directory} directory for token file.");
+    }
   }
 
   /**
    * {@inheritdoc}
    */
   public function removeToken(): void {
-    // Remove the token data from storage.
+    // Removes the token data from the file without removing the file.
     $this->saveToken([]);
   }
 
   /**
-   * Gets the storage location for OAuth token data.
-   *
-   * TODO DI the config factory, save the location (directory only) to a
-   * class property. Only append the file name to the path in saveToken().
+   * Removes the file in which the OAuth token data is stored.
    */
-  public static function oauthTokenPath() {
-    static $token_path;
-
-    if (!isset($token_path)) {
-      // Get the file location from config.
-      $location = \Drupal::config(AuthenticationForm::CONFIG_NAME)->get('oauth_token_storage_location');
-
-      $token_path = (!empty($location) ? rtrim($location, " \t\n\r\0\x0B\\/") : 'private://.apigee_edge') . '/oauth.dat';
+  public function removeTokenFile(): void {
```

Comment:
Should this be added to the interface?

Reply:
It should not, because there is no `OauthTokenFileStorageInterface`, only a `OauthTokenStorageInterface` which defines the basic requirements of an Oauth token store service implementations but it does not "care" about the actual implementation of them. (How and where they store the token data - removeTokenFile() is related to _how_ the data is being stored.)
This concrete implementation stores the Oauth token data in a file, a custom implementation can store it in a config object and the implementation could have a public `removeConfigObject()`.